fix(composer): do not use `--minimal-changes` with `lockFileMaintenance` and add `postUpdateOption` for not using it

elaman opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes

<!-- Describe what behavior is changed by this PR. -->
Currently `--minimal-changes` flag is being applied to `composer` commands. It was recently introduced in PR #34218. This has caused following issue #39823. As a solution, new `composerMinimalChanges` configuration is introduced, which allows `--minimal-changes` to be optionally applied instead of being a default behavior. 

## Context

Please select one of the following:

- [x] This closes an existing Issue, Closes: #39823 <!-- NOTE that this should NOT be a Discussion -->
- [ ] This doesn't close an Issue, but I accept the risk that this PR may be closed if maintainers disagree with its opening or implementation

## AI assistance disclosure

<!-- We request this information to assist reviewers in identifying AI-generated errors and other issues specific to AI usage. While we typically permit the use of AI tools, we appreciate being notified when they are employed. -->

Did you use AI tools to create any part of this pull request?

Please select one option and, if yes, briefly describe how AI was used (e.g., code, tests, docs) and which tool(s) you used.

- [x] No — I did not use AI for this contribution.
- [ ] Yes — minimal assistance (e.g., IDE autocomplete, small code completions, grammar fixes).
- [ ] Yes — substantive assistance (AI-generated non‑trivial portions of code, tests, or documentation).
- [ ] Yes — other (please describe):

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests, but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

The public repository: <URL>

<!-- If you have any suggestions about this PR template, edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. You can commit as many times as you need in this branch. -->

elaman committed:
Make --minimal-config configurable.

elaman committed:
Fix accidental remove.

elaman committed:
Remove composerMinimalChange from wrong file

RahulGautamSingh reviewed (changes requested):
Add the type to `RenovateConfig` instead .. it should get passed down from there where needed.

elaman committed:
Update docs/usage/configuration-options.md

Co-authored-by: RahulGautamSingh <rahultesnik@gmail.com>

elaman committed:
Add composerMinimalChanges to the renovateconfig interface.

talkinnl commented:
> Thank you for opening this PR. I was just researching why renovate stopped creating the desired lock file maintenance and found this exact flag being used in the log. For completeness, composer even notifies about this being a mistake in this command.
> 
> ```
> The --minimal-changes option should be used with package arguments or after modifying composer.json requirements, otherwise it will likely not yield any dependency changes.
> ```
> 
> However, I do not think it should be made configurable in the way you suggest. It is great to use with the actual update commands, but it should never be used with the lock file maintance action.

I also don't think it should be a config option. But even if it was, I think the old behavior should be the default.

The way the flag was added recently just completely disables using 'Lock file maintenance' for security updates of indirect dependencies for PHP.

viceice reviewed (changes requested):
I think the simpler solution is to delete the lockfile before running lockfile maintenance like we do on allmost all other managers. 🤔

@jamietanna @secustor wdyt?

elaman commented:
@viceice and someone else please correct me if I'm wrong, but removing lock file would change the "content hash" resulting in a new PR even if there are no updates. While perhaps not bad, this is different from the behavior Renovate had before this issue. My sole aim here is/was to run composer update without that flag, as it does very good job as is without removing a lock file. That being said I'm open to any quick solution that will restore lock file maintenances ability to apply updates (sometimes security related) of unspecified dependencies on schedule.

elaman committed:
Use postupdateoptions.

elaman committed:
Remove the lock file before lock file maintenance upgrade.

elaman committed:
Import

elaman commented:
@talkinnl good to know, thanks. @viceice I've moved the flag to be used only via postUpdateOptions, which applies for regular updates only. Also added the lock file removal.

talkinnl commented on docs/usage/configuration-options.md:
```diff
@@ -3848,6 +3848,7 @@ Table with options:
 | ---------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------- |
 | `bundlerConservative`        | Enable conservative mode for `bundler` (Ruby dependencies). This will only update the immediate dependency in the lockfile instead of all subdependencies. |
 | `composerWithAll`            | Run `composer update` with `--with-all-dependencies` flag instead of the default `--with-dependencies`.                                                    |
+| `composerMinimalChanges`     | Run `composer update` with `--minimal-changes` flag for (except for lock file maintanance).                                                                |
```
Yes, it's a restore. The original PR was the BC change, and in hindsight that PR should really not have gotten away with 'no documentation needed' too.

elaman replied:
@jamietanna understood. Can we at least keep the variable name `composerMinimalChanges` and reverse the default value to `true` at least?

elaman committed:
Merge branch 'main' into make-composer-minimal-changes-optional

jamietanna commented on docs/usage/configuration-options.md:
```diff
@@ -3849,6 +3849,7 @@ Table with options:
 | ---------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------- |
 | `bundlerConservative`        | Enable conservative mode for `bundler` (Ruby dependencies). This will only update the immediate dependency in the lockfile instead of all subdependencies. |
 | `composerWithAll`            | Run `composer update` with `--with-all-dependencies` flag instead of the default `--with-dependencies`.                                                    |
+| `composerMinimalChanges`     | Run `composer update` with `--minimal-changes` flag for (except for lock file maintanance).                                                                |
```
```suggestion
| `composerMinimalChanges`     | Run `composer update` with `--minimal-changes` flag for (except for lock file maintenance).                                                                |
```

elaman committed:
Flip to use composerNoMinimalChanges

elaman committed:
Minor fixes.

viceice commented:
missing code coverage

elaman committed:
Update artifacts.spec.ts

elaman committed:
Ignore tests for exception during lock file deletion (similar to npm, pnpm, etc)

elaman committed:
Move v8 comment.

elaman committed:
Merge branch 'renovatebot:main' into make-composer-minimal-changes-optional

elaman committed:
Merge branch 'main' into make-composer-minimal-changes-optional

jamietanna commented on lib/modules/manager/composer/artifacts.ts:
```diff
@@ -172,6 +173,21 @@ export async function updateArtifacts({
       commands.push('git stash pop || true');
     }
 
+    if (config.isLockFileMaintenance) {
```
Yeah I'd prefer it to be separate, please - thanks!

elaman committed:
Remove composer lock removal

jamietanna commented on docs/usage/configuration-options.md:
```diff
@@ -3885,6 +3885,7 @@ Table with options:
 | ---------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------- |
 | `bundlerConservative`        | Enable conservative mode for `bundler` (Ruby dependencies). This will only update the immediate dependency in the lockfile instead of all subdependencies. |
 | `composerWithAll`            | Run `composer update` with `--with-all-dependencies` flag instead of the default `--with-dependencies`.                                                    |
+| `composerNoMinimalChanges`   | Run `composer update` with no `--minimal-changes` flag (except for lock file maintenance, where it is a default behavior).                                 |
```
```suggestion
| `composerNoMinimalChanges`   | Run `composer update` with no `--minimal-changes` flag (does not affect lock file maintenance, which will never use `--minimal-changes`).            |
```

jamietanna committed:
docs(composer): reword `composerNoMinimalChanges`

jamietanna committed:
Fix Markdown formatting

jamietanna reviewed (approved):
I think given there are two things we're doing, we could've split this further into a `fix: don't use with lockFileMaintenance` and a `feat(...): add postupdateOption` but this is fine 👍
